Allow detecting and setting media rotation

g7 opened the pull request:
Fixes #310

Rotation is done through GskTransforms directly on the Snapshot. 

The default behaviour is to autodetect the media rotation if available, but frontends can force it by setting the `rotation` property.

I've also exposed this to the actual app, but I'm unsure if that's wanted :sweat_smile:

This is how it looks now (playing the video from the aforementioned bug report):

[Video del 2023-04-01 15-36-57.webm](https://user-images.githubusercontent.com/92799/229292631-562b4cf8-ff7b-421b-8e45-c389e838c9c5.webm)

Thanks for your work on clapper!

Rafostar commented:
Thanks for working on this 👍 

> Rotation is done through GskTransforms directly on the Snapshot.

Yes, this is the correct behaviour I had in mind for this feature.

Rafostar commented on lib/gst/plugin/gstclappersink.c:
```diff
@@ -54,6 +56,29 @@ G_DEFINE_TYPE_WITH_CODE (GstClapperSink, gst_clapper_sink, GST_TYPE_VIDEO_SINK,
 GST_ELEMENT_REGISTER_DEFINE (clappersink, "clappersink", GST_RANK_NONE,
     GST_TYPE_CLAPPER_SINK);
 
+#define GST_TYPE_CLAPPER_SINK_ROTATION (gst_clapper_sink_rotation_get_type())
+static GType
+gst_clapper_sink_rotation_get_type (void)
+{
+  static GType rotation_type = 0;
+
+  static const GEnumValue rotations[] = {
+    {GST_VIDEO_ORIENTATION_AUTO, "Auto", "automatic"},
+    {GST_VIDEO_ORIENTATION_IDENTITY, "Original", "none"},
+    {GST_VIDEO_ORIENTATION_90R, "90°", "clockwise"},
+    {GST_VIDEO_ORIENTATION_180, "180°", "rotate-180"},
+    {GST_VIDEO_ORIENTATION_90L, "270°", "counterclockwise"},
+    {0, NULL, NULL},
+  };
+
+  if (!rotation_type) {
+    rotation_type = g_enum_register_static ("GstClapperSinkRotation",
+        rotations);
+  }
+  return rotation_type;
+}
+
+
```
This seems unnecessary, considering a `GST_TYPE_VIDEO_ORIENTATION_METHOD` type already exists since GStreamer 1.20. Maybe we should just use that for consistency in values? It should be (I hope) relatively easy to implement remaining ones with `gsk_transform_rotate_3d` probably? And do 180 rotation around X/Y axis for flip? Or am I mistaken here? 🤔

Rafostar commented on lib/gst/plugin/gstclappersink.c:
```diff
@@ -869,6 +940,12 @@ gst_clapper_sink_class_init (GstClapperSinkClass *klass)
           DEFAULT_KEEP_LAST_FRAME,
           G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS));
 
+  g_object_class_install_property (gobject_class, PROP_ROTATION,
+      g_param_spec_enum ("rotation", "Video rotation",
```
Maybe lets call this "rotate-method" with `PROP_ROTATE_METHOD` for consistency with other GStreamer elements. Example: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/blob/7755cd2f7a40e92f35d50765af09279384ec9a0e/subprojects/gst-plugins-bad/sys/d3d11/gstd3d11videosink.cpp#L343

Rafostar commented on lib/gst/plugin/gstclappersink.c:
```diff
@@ -869,6 +940,12 @@ gst_clapper_sink_class_init (GstClapperSinkClass *klass)
+  g_object_class_install_property (gobject_class, PROP_ROTATION,
+      g_param_spec_enum ("rotation", "Video rotation",
+          "The video rotation",
+          GST_TYPE_CLAPPER_SINK_ROTATION, DEFAULT_ROTATION,
```
`GST_TYPE_VIDEO_ORIENTATION_METHOD` ?

Rafostar commented on lib/gst/plugin/gstclappersink.c:
```diff
@@ -889,6 +967,8 @@ gst_clapper_sink_class_init (GstClapperSinkClass *klass)
 
   sink_pad_templ = gst_clapper_importer_loader_make_sink_pad_template ();
   gst_element_class_add_pad_template (gstelement_class, sink_pad_templ);
+
+  gst_type_mark_as_plugin_api (GST_TYPE_CLAPPER_SINK_ROTATION, 0);
```
If we change to `GST_TYPE_VIDEO_ORIENTATION_METHOD` this won't be needed anymore.

Rafostar commented on lib/gst/plugin/gstclappersink.c:
```diff
@@ -633,6 +669,31 @@ gst_clapper_sink_stop (GstBaseSink *bsink)
+static gboolean
+gst_clapper_sink_event (GstBaseSink *bsink, GstEvent *event)
+{
+  GstClapperSink *self = GST_CLAPPER_SINK (bsink);
```
```suggestion
  GstClapperSink *self = GST_CLAPPER_SINK_CAST (bsink);
```

Rafostar commented on lib/gst/plugin/gstclappersink.c:
```diff
@@ -643,6 +704,15 @@ gst_clapper_sink_change_state (GstElement *element, GstStateChange transition)
 
   switch (transition) {
+    case GST_STATE_CHANGE_NULL_TO_READY:
+      /* Reset stream_orientation */
+      self->stream_orientation = GST_VIDEO_ORIENTATION_IDENTITY;
+      if (self->rotation_mode == GST_VIDEO_ORIENTATION_AUTO) {
+        GST_CLAPPER_SINK_LOCK (self);
+        gst_clapper_paintable_set_rotation (self->paintable, self->stream_orientation);
+        GST_CLAPPER_SINK_UNLOCK (self);
+      }
```
This whole block should probably be within locked mutex (not only set_rotation).

Rafostar commented on lib/gst/plugin/gstclappersink.c:
```diff
@@ -633,6 +669,31 @@ gst_clapper_sink_stop (GstBaseSink *bsink)
+    gst_event_parse_tag (event, &taglist);
+
+    if (gst_video_orientation_from_tag (taglist, &orientation)) {
+      self->stream_orientation = orientation;
+
+      if (self->rotation_mode == GST_VIDEO_ORIENTATION_AUTO)
+        gst_clapper_paintable_set_rotation (self->paintable, orientation);
+    }
```
Missing `GST_CLAPPER_SINK_LOCK` for editing/reading "self" values.

Rafostar commented:
> I've also exposed this to the actual app, but I'm unsure if that's wanted

As for adding this to the GUI, I would probably prefer to not clutter it too much with buttons for rarely needed/used option. The "auto" behaviour should already fix the issue you linked. For the rest of cases where it should not be rotated, just user wishes to do so for some reason, I would probably simply put a keyboard shortcuts for +/- 90 degrees changes. Does not have to be done as part of this MR.

If this is alright with you, I would keep it for now for testing purposes of this MR and remove before merging.

g7 commented:
Thanks for the review.

> If this is alright with you, I would keep it for now for testing purposes of this MR and remove before merging.

Sure.

Updated the pull request with changes addressing the raised issues, including the implementation of the missing methods. I've added the commits on top rather than squashing directly to ease the review.

The reason why I added a custom enum was to avoid cluttering the UI side with too *many* options, so I kept only the ones that made sense to me. But obviously since we're going to get rid of that part altogether that's not a worry anymore.

Rafostar commented on lib/gst/plugin/gstgtkutils.c:
```diff
@@ -146,7 +146,10 @@ gst_gtk_get_width_height_for_rotation (gint width, gint height,
+  if (rotation == GST_VIDEO_ORIENTATION_90R ||
+      rotation == GST_VIDEO_ORIENTATION_90L ||
+      rotation == GST_VIDEO_ORIENTATION_UL_LR ||
+      rotation == GST_VIDEO_ORIENTATION_UR_LL) {
```
nitpick: With so many, a switch-case might be cleaner here for setting `out_width` and `out_height`.

g7 commented:
v3:

use a switch in `gst_gtk_get_width_height_for_rotation`, uniformise the indentation in `gst_clapper_paintable_snapshot_internal` with the rest of the code

Rafostar commented:
Seems OK to me. Also tired if mouse navigation coordinates are correct with this and they seem to be. Good job :+1: 

In your free time please clean commits of this, removing changes to the UI as we discussed earlier and we will be able to get this in.

g7 commented:
Thanks!

v4:
- Squashed review commits
- Removed UI bits
- Uniformised indentation of the switch-case in `gst_clapper_sink_event`, missed that before
